Print main parameters while configuring the wrapper

lfoppiano opened the pull request:
This PR add the printing of certain parameters. I'm not sure it's appropriate to place it into the applications, rather than in the wrapper, which would be more uniformed. 

The format is something like: 

```
---
max_epoch: 60
batch_size: 80
max_sequence_length: 30
model_name: grobid-date-BERT_CRF
---
```

lfoppiano commented:
Here an example (I commented out the trainining to save time). Notice that it's disabled completely for the evaluation. 

- train_eval with fold-count > 1

```
/Users/lfoppiano/opt/anaconda3/envs/delft2/bin/python -m delft.applications.grobidTagger --architecture BERT_CRF --transformer allenai/scibert_scivocab_cased/dir date train_eval --fold-count 2
Loading data...
514 train sequences
58 validation sequences
64 evaluation sequences
---
max_epoch: 60
early_stop: True
batch_size: 80
max_sequence_length: 30
model_name: grobid-date-BERT_CRF
---
Output directory: data/models/sequenceLabelling/grobid-date-BERT_CRF

------------------------ fold 0--------------------------------------
allenai/scibert_scivocab_cased/dir will be used, loaded via local_model_dir
Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
==================================================================================================
 input_token (InputLayer)       [(None, None)]       0           []                               
                                                                                                  
 input_attention_mask (InputLay  [(None, None)]      0           []                               
 er)                                                                                              
                                                                                                  
 input_token_type (InputLayer)  [(None, None)]       0           []                               
                                                                                                  
 tf_bert_model (TFBertModel)    TFBaseModelOutputWi  109938432   ['input_token[0][0]',            
                                thPoolingAndCrossAt               'input_attention_mask[0][0]',   
                                tentions(last_hidde               'input_token_type[0][0]']       
                                n_state=(None, None                                               
                                , 768),                                                           
                                 pooler_output=(Non                                               
                                e, 768),                                                          
                                 past_key_values=No                                               
                                ne, hidden_states=N                                               
                                one, attentions=Non                                               
                                e, cross_attentions                                               
                                =None)                                                            
                                                                                                  
 dropout_37 (Dropout)           (None, None, 768)    0           ['tf_bert_model[0][0]']          
                                                                                                  
==================================================================================================
Total params: 109,938,432
Trainable params: 109,938,432
Non-trainable params: 0
__________________________________________________________________________________________________
Model: "crf_model_wrapper_for_bert"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 crf (CRF)                   multiple                  5446      
                                                                 
 model (Functional)          (None, None, 768)         109938432 
                                                                 
=================================================================
Total params: 109,943,878
Trainable params: 109,943,878
Non-trainable params: 0
_________________________________________________________________

------------------------ fold 1--------------------------------------
allenai/scibert_scivocab_cased/dir will be used, loaded via local_model_dir
training runtime: 13.865 seconds 

Evaluation:

------------------------ fold 0 --------------------------------------
allenai/scibert_scivocab_cased/dir will be used, loaded via delft_model
loading model weights data/models/sequenceLabelling/grobid-date-BERT_CRF/model_weights0.hdf5
	f1 (micro): 36.46
                  precision    recall  f1-score   support

           <day>     0.3115    0.4524    0.3689        42
         <month>     0.0000    0.0000    0.0000        59
          <year>     0.3500    0.7656    0.4804        64

all (micro avg.)     0.3383    0.4121    0.3716       165


------------------------ fold 1 --------------------------------------
allenai/scibert_scivocab_cased/dir will be used, loaded via delft_model
loading model weights data/models/sequenceLabelling/grobid-date-BERT_CRF/model_weights1.hdf5
	f1 (micro): 10.34
                  precision    recall  f1-score   support

           <day>     0.2581    0.1905    0.2192        42
         <month>     0.2857    0.0678    0.1096        59
          <year>     0.0435    0.0469    0.0451        64

all (micro avg.)     0.1316    0.0909    0.1075       165

----------------------------------------------------------------------

** Worst ** model scores - run 1
                  precision    recall  f1-score   support

           <day>     0.2581    0.1905    0.2192        42
         <month>     0.2857    0.0678    0.1096        59
          <year>     0.0435    0.0469    0.0451        64

all (micro avg.)     0.1316    0.0909    0.1075       165


** Best ** model scores - run 0
                  precision    recall  f1-score   support

           <day>     0.3115    0.4524    0.3689        42
         <month>     0.0000    0.0000    0.0000        59
          <year>     0.3500    0.7656    0.4804        64

all (micro avg.)     0.3383    0.4121    0.3716       165

loading model weights data/models/sequenceLabelling/grobid-date-BERT_CRF/model_weights0.hdf5
----------------------------------------------------------------------

Average over 2 folds
                  precision    recall  f1-score   support

           <day>     0.2848    0.3214    0.2941        42
         <month>     0.1429    0.0339    0.0548        59
          <year>     0.1967    0.4062    0.2628        64

all (micro avg.)     0.2235    0.2515    0.2340          

model config file saved
preprocessor saved
transformer config saved
transformer tokenizer saved
model saved

Process finished with exit code 0

```

kermitt2 commented:
For the summary, what about just adding the following in `BaseModel`:

```python
def print_summary(self):
      self.model.summary()
```

instead of modifying all the methods?

For CRF tf-addons base model, as we need to print both inner model and CRF wrapper:

```python
def print_summary(self):
      seld.model.base_model.summary()
      self.model.summary()
```

more compact and readable?

kermitt2 commented on delft/sequenceLabelling/wrapper.py:
```diff
@@ -134,6 +134,14 @@ def __init__(self,
+        print("max_sequence_length:", max_sequence_length)
+        print("model_name:", model_name)
+        print("---")
+
```
I would put it in the train/eval method? Because these values are not the correct ones when the model will be then loaded, eg:

```
        model = Sequence(model_name)
        model.load()
```

It's only after `load()` than the config values are set.

When training from scratch, the config values could also be changed between the model creation and the actual training, although it's never the case in the current `applications/*` tasks.

lfoppiano replied:
Where do you mean to move the `prints`? In the `application/*` ?

kermitt2 replied:
Rather in the wrapper.train() and/or wrapper.eval(), because we know at this point that the configuration parameters are set (this is not the case when instantiating the architecture) and I guess that printing parameters is relevant only for train and eval.

kermitt2 commented on delft/sequenceLabelling/wrapper.py:
```diff
@@ -134,6 +134,14 @@ def __init__(self,
+        print("max_sequence_length:", max_sequence_length)
+        print("model_name:", model_name)
+        print("---")
+
```
We might want to add the `learning_rate` and `use-ELMo`? Because they are parameters that can impact quite a lot a training.

kermitt2 commented on delft/textClassification/models.py:
```diff
@@ -285,7 +284,7 @@ def train_folds(X, y, model_config, training_config, embeddings, callbacks=None)
         val_y = y[fold_start:fold_end]
 
-        foldModel = getModel(model_config, training_config)
+        foldModel = getModel(model_config, training_config, print_summary=fold_id == 0)
```
I find it hard to read... probably pythonesque and subjective, but I really don't like one-liner for maintenance and when taking over old code.

kermitt2 commented:
When having this

```python
def print_summary(self):
      self.model.summary()
```

I think the idea is then to simply call it when needed so in the `wrapper.py` after the `get_model` call.

```
self.model = get_model(self.model_config, self.p, len(self.p.vocab_tag), load_pretrained_weights=True)
self.model.print_summary()
```

So no need for propagating all the `print_summary` boolean parameters everywhere in the models.py.

lfoppiano commented:
I modified the print and reverted some of the initial complicated logic. I also included the evaluation which could be useful to include when publishing results. 
I did not test the classification part, though

lfoppiano commented:
When I try to test the classification, I have a problem with circular dependencies Transformer - BertPreprocessor. This can be solved by removing the import from delft/sequenceLabelling/__init__ which is invoked when importing the BErtPreprocessor from Transformer. I'm not sure how else to fix it.. 

Once I fix that I try to run the citationClassifier and I get this: 

```
(delft2) Lucas-MacBook-Pro:delft lfoppiano$ python -m delft.applications.citationClassifier train
loading citation sentiment corpus...
---
max_epoch: 60
early_stop: True
batch_size: 256
max_sequence_length: 150
model_name: citations_gru
learning_rate:  0.001
use_ELMo:  None
---
Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
==================================================================================================
 input_1 (InputLayer)           [(None, 150, 300)]   0           []                               
                                                                                                  
 bidirectional (Bidirectional)  (None, 150, 128)     140544      ['input_1[0][0]']                
                                                                                                  
 dropout (Dropout)              (None, 150, 128)     0           ['bidirectional[0][0]']          
                                                                                                  
 bidirectional_1 (Bidirectional  (None, 150, 128)    74496       ['dropout[0][0]']                
 )                                                                                                
                                                                                                  
 global_max_pooling1d (GlobalMa  (None, 128)         0           ['bidirectional_1[0][0]']        
 xPooling1D)                                                                                      
                                                                                                  
 global_average_pooling1d (Glob  (None, 128)         0           ['bidirectional_1[0][0]']        
 alAveragePooling1D)                                                                              
                                                                                                  
 concatenate (Concatenate)      (None, 256)          0           ['global_max_pooling1d[0][0]',   
                                                                  'global_average_pooling1d[0][0]'
                                                                 ]                                
                                                                                                  
 dense (Dense)                  (None, 32)           8224        ['concatenate[0][0]']            
                                                                                                  
 dense_1 (Dense)                (None, 3)            99          ['dense[0][0]']                  
                                                                                                  
==================================================================================================
Total params: 223,363
Trainable params: 223,363
Non-trainable params: 0
__________________________________________________________________________________________________
Traceback (most recent call last):
  File "/Users/lfoppiano/opt/anaconda3/envs/delft2/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/lfoppiano/opt/anaconda3/envs/delft2/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/lfoppiano/development/projects/delft/delft/applications/citationClassifier.py", line 142, in <module>
    train(embeddings_name, args.fold_count, architecture=architecture, transformer=transformer)
  File "/Users/lfoppiano/development/projects/delft/delft/applications/citationClassifier.py", line 48, in train
    model.train(xtr, y)
  File "/Users/lfoppiano/development/projects/delft/delft/textClassification/wrapper.py", line 161, in train
    self.model.train_model(
  File "/Users/lfoppiano/development/projects/delft/delft/textClassification/models.py", line 149, in train_model
    loss = self.model.fit(
  File "/Users/lfoppiano/opt/anaconda3/envs/delft2/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 67, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/Users/lfoppiano/development/projects/delft/delft/textClassification/data_generator.py", line 46, in __getitem__
    batch_x, batch_y = self.__data_generation(index)
  File "/Users/lfoppiano/development/projects/delft/delft/textClassification/data_generator.py", line 80, in __data_generation
    input_ids, input_masks, input_segments = create_batch_input_bert(self.x[(index*self.batch_size):(index*self.batch_size)+max_iter], 
  File "/Users/lfoppiano/development/projects/delft/delft/textClassification/preprocess.py", line 63, in create_batch_input_bert
    encoded_tokens = transformer_tokenizer.batch_encode_plus(texts, add_special_tokens=True, truncation=True, 
AttributeError: 'NoneType' object has no attribute 'batch_encode_plus'

```

I have the same issue with `softwareClassifier`

lfoppiano commented:
Sorry, I commented the training to speed up the testing and I forgot to put it back. It's now back.

kermitt2 commented:
There is no place to add a comment in these files, but I think `data/db/.gitkeep` and `data/download/.gitkeep` must stay because they are not created by the existing code and it will result in failing command line from a fresh install.

kermitt2 commented on delft/textClassification/wrapper.py:
```diff
@@ -226,28 +228,31 @@ def predict(self, texts, output_format='json', use_main_thread_only=False):
 
-                result = self.model.predict(test_generator, use_main_thread_only=use_main_thread_only)
-            else:
-                raise (OSError('Could not find a model.'))
```
Any reason to have removed this conditional? It was to inform the user in case he was launching an eval without a model or n models saved (or if he makes a mismatched between simple model and n models).

lfoppiano replied:
I think the check for the model was moved at the beginning of the method to simplify the ifs

kermitt2 replied:
Only the check on model has been moved, not the check on models (case n fold > 1).

lfoppiano replied:
ahh true... let me try to fix it
